File: Develop/frontend/src/types/systemFunctions.ts
/**
 * System Functions Types
 * 系統功能設定相關類型定義（正名化版本）
 */

export interface SystemFunction {
  id: number;
  func_code: string;
  upper_func_id: number;
  func_cname: string;
  func_ename: string;
  func_type: number; // 1:節點/選單, 2:功能
  func_order: number;
  func_icon?: string;
  module_code?: string; // 正名化欄位（原 func_module_name）
  module_item: string[];
  description?: string;
  is_mana: boolean;
  is_active: boolean;
  edit_by: number;
  created_at: string;
  updated_at?: string;
  children?: SystemFunction[]; // 樹狀結構用
}

export interface SystemFunctionCreate {
  func_code: string;
  upper_func_id: number;
  func_cname: string;
  func_ename: string;
  func_type: number;
  func_order: number;
  func_icon?: string;
  module_code?: string; // 正名化欄位
  module_item?: string[];
  description?: string;
  is_mana?: boolean;
  is_active?: boolean;
}

export interface SystemFunctionUpdate {
  func_code?: string;
  upper_func_id?: number;
  func_cname?: string;
  func_ename?: string;
  func_type?: number;
  func_order?: number;
  func_icon?: string;
  module_code?: string; // 正名化欄位
  module_item?: string[];
  description?: string;
  is_mana?: boolean;
  is_active?: boolean;
}

export interface GetSystemFunctionsParams {
  skip?: number;
  limit?: number;
  is_active?: boolean;
  func_type?: number;
  search?: string;
}

File: Develop/backend/app/schemas/system_functions.py
"""
System Function Schemas
系統功能設定表 API Schema（正名化版本）
"""

from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict


class SystemFunctionBase(BaseModel):
    """系統功能基本資料"""
    func_code: str = Field(..., max_length=200, description="功能代碼，用於權限識別、前端路由、日誌記錄")
    upper_func_id: int = Field(0, description="上層功能ID (0表示根節點)")
    func_cname: str = Field(..., max_length=200, description="功能中文名稱")
    func_ename: str = Field(..., max_length=200, description="功能英文名稱")
    func_type: int = Field(..., description="功能類型 (1:節點/選單, 2:功能)")
    func_order: int = Field(..., description="排序順序")
    func_icon: Optional[str] = Field(None, max_length=200, description="圖示")
    module_code: Optional[str] = Field(None, max_length=200, description="模組代碼，用於 API 路由識別")
    module_item: List[str] = Field(default_factory=list, description="可設定權限項目 (create/read/update/delete/print/file)")
    description: Optional[str] = Field(None, description="功能說明")
    is_mana: bool = Field(False, description="是否為管理功能")
    is_active: bool = Field(True, description="是否啟用")


class SystemFunctionCreate(SystemFunctionBase):
    """建立系統功能"""
    pass


class SystemFunctionUpdate(BaseModel):
    """更新系統功能（所有欄位可選）"""
    func_code: Optional[str] = Field(None, max_length=200, description="功能代碼")
    upper_func_id: Optional[int] = Field(None, description="上層功能ID")
    func_cname: Optional[str] = Field(None, max_length=200, description="功能中文名稱")
    func_ename: Optional[str] = Field(None, max_length=200, description="功能英文名稱")
    func_type: Optional[int] = Field(None, description="功能類型")
    func_order: Optional[int] = Field(None, description="排序順序")
    func_icon: Optional[str] = Field(None, max_length=200, description="圖示")
    module_code: Optional[str] = Field(None, max_length=200, description="模組代碼")
    module_item: Optional[List[str]] = Field(None, description="可設定權限項目")
    description: Optional[str] = Field(None, description="功能說明")
    is_mana: Optional[bool] = Field(None, description="是否為管理功能")
    is_active: Optional[bool] = Field(None, description="是否啟用")


class SystemFunctionResponse(SystemFunctionBase):
    """系統功能回應資料"""
    id: int
    edit_by: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class SystemFunctionTreeNode(SystemFunctionResponse):
    """系統功能樹狀結構節點（含子節點）"""
    children: Optional[List['SystemFunctionTreeNode']] = Field(default_factory=list, description="子功能列表")

    model_config = ConfigDict(from_attributes=True)


# 解決循環引用問題
SystemFunctionTreeNode.model_rebuild()
